File: 86_binary-search-tree-iterator/binary-search-tree-iterator.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/binary-search-tree-iterator
@Language: C++
@Datetime: 14-11-14 04:00
*/

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 * Example of iterate a tree:
 * Solution iterator = Solution(root);
 * while (iterator.hasNext()) {
 *    TreeNode * node = iterator.next();
 *    do something for node
 */
class Solution {
public:
    //@param root: The root of binary tree.
    Solution(TreeNode *root) {
        // write your code here
        inOrder(root);
    }

    //@return: True if there has next node, or false
    bool hasNext() {
        // write your code here
        return !q.empty();
    }
    
    //@return: return next node
    TreeNode* next() {
        // write your code here
        auto item = q.front();
        q.pop();
        return item;
    }
private:
    queue<TreeNode*> q;
    void inOrder(TreeNode* root)
    {
        if(root==NULL) return;
        inOrder(root->left);
        q.push(root);
        inOrder(root->right);
    }
};

File: 86_binary-search-tree-iterator/binary-search-tree-iterator.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/binary-search-tree-iterator
@Language: Python
@Datetime: 14-12-25 08:52
'''

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None

Example of iterate a tree:
iterator = Solution(root)
while iterator.hasNext():
    node = iterator.next()
    do something for node 
"""
class Solution:
    cur, it = None, None
    #@param root: The root of binary tree.
    def __init__(self, root):
        self.it = self.inOrder(root)

    #@return: True if there has next node, or false
    def hasNext(self):
        try:
            self.cur = self.it.next()
            return True
        except:
            return False
        
    #@return: return next node
    def next(self):
        return self.cur
    
    def inOrder(self, root):
        if root == None: return;
        for x in self.inOrder(root.left): yield x
        yield root
        for x in self.inOrder(root.right): yield x
    
